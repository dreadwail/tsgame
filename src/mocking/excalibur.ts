import { Engine, Logger, Scene } from 'excalibur';

export const createMockLogger = (): Logger => {
  const mock = {
    error: jest.fn(),
    warn: jest.fn(),
  };
  return (mock as any) as Logger;
};

export const createMockEngine = (): Engine => {
  const mock = {
    drawWidth: 800,
    drawHeight: 600,
    halfDrawWidth: 400,
    halfDrawHeight: 300,
    add: jest.fn((...args: any[]) => {
      if (args[1] instanceof Scene) {
        const scene = args[1];
        scene.engine = (mock as any) as Engine;
      }
    }),
    start: jest.fn(() => ({
      then: jest.fn(),
    })),
    goToScene: jest.fn(),
  };
  return (mock as any) as Engine;
};
