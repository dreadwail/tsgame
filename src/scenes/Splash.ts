import { Color, Engine, Label, Scene, TextAlign, Vector } from 'excalibur';
import { ILabelArgs } from 'excalibur/dist/Label';

const baseLabelOptions: ILabelArgs = {
  fontFamily: 'Sigmar One',
  textAlign: TextAlign.Center,
};

export default class Splash extends Scene {
  public onInitialize(engine: Engine) {
    const gameName = this.createGameName();
    const pressKey = this.createPressAnyKeyToContinue();
    const credit = this.createCredit();

    gameName.pos = new Vector(engine.halfDrawWidth, engine.halfDrawHeight);
    pressKey.pos = new Vector(engine.halfDrawWidth, gameName.getBottom());
    credit.pos = new Vector(engine.halfDrawWidth, engine.drawHeight - 70);

    this.add(gameName);
    this.add(pressKey);
    this.add(credit);
  }

  private createGameName = (): Label => {
    return new Label({
      ...baseLabelOptions,
      text: 'Invaded',
      fontSize: 150,
      color: Color.Red,
    });
  };

  private createPressAnyKeyToContinue = (): Label => {
    return new Label({
      ...baseLabelOptions,
      text: '(press any key to continue)',
      fontSize: 35,
      color: Color.White,
    });
  };

  private createCredit = (): Label => {
    return new Label({
      ...baseLabelOptions,
      text: 'Â© Ben Lakey 2018',
      fontSize: 20,
      color: Color.Gray,
    });
  };
}
