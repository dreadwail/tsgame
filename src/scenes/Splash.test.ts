import { Actor, Engine, Label } from 'excalibur';

import { createMockEngine } from '../mocking/excalibur';

import Splash from './Splash';

describe('Splash', () => {
  let engine: Engine;
  let splash: Splash;

  beforeEach(() => {
    engine = createMockEngine();
    splash = new Splash(engine);
    engine.add(Splash.sceneName, splash);
  });

  describe('onInitialize', () => {
    beforeEach(() => {
      splash.onInitialize(engine);
    });

    it('initializes the correct labels', () => {
      const labels = splash.actors
        .filter((actor: Actor): actor is Label => actor instanceof Label)
        .map(label => label.text);

      expect(labels).toEqual(['Invaded', '(press any key to continue)', 'Â© Ben Lakey 2018']);
    });
  });

  describe('onActivate', () => {
    beforeEach(() => {
      splash.onInitialize(engine);
      splash.onActivate();
    });

    it('correctly positions the menu entries', () => {
      const labels = splash.actors
        .filter((actor: Actor): actor is Label => actor instanceof Label)
        .map(label => label.pos);

      expect(labels).toMatchSnapshot();
    });
  });
});
