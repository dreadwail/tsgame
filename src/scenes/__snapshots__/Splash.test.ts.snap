// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splash onActivate correctly positions the menu entries 1`] = `
Array [
  Vector {
    "x": 400,
    "y": 300,
  },
  Vector {
    "x": 400,
    "y": 375,
  },
  Vector {
    "x": 400,
    "y": 580,
  },
]
`;
